### PR TITLE
fix: to avoid dead lock, always put txhashset.write before store.batch

### PR DESCRIPTION
**Pull request by garyyu**
Fix for https://github.com/mimblewimble/grin/issues/1698

And we have to always remember in the future that `txhashset.write()` must always before `store.batch()`.

Perhaps somebody has better fix solution for this?  wrap a dedication function? or define a macro?

### CODE REVIEW

**Comment by garyyu** [comment]
Note: just find this PR can fix this problem, but it will have dead lock for `TxHashsetSave`, I'm investigating on it.

**Comment by garyyu** [comment]
Find it:
```
	pub fn txhashset_write(
		...
		let mut batch = self.store.batch()?;
			...
			let mut txhashset_ref = self.txhashset.write().unwrap();
```

@antiochp  May I move this `txhashset.write()` after  `batch.commit()?;`?    will update the code and request review.

**Comment by garyyu** [comment]
Tested by 2 times of the whole fast sync procedure, and with API call (by following script) on every 0.3s, no more dead lock issue.

```
#!/bin/bash

interval=0.3

if [ "$1" != "" ]; then
    interval=$1
fi

SECONDS=0
while true
do
    variableA=$(curl -0 -XGET  http://127.0.0.1:13413/v1/blocks/128000 2>null | grep block_height | tr -d ,)
    printf "`date '+%Y-%m-%d %H:%M:%S'`  $variableA\n"
    sleep $interval
done
```

**Comment by NgoziDim** [comment]
smh

File: chain/src/chain.rs [hunk 1]
```diff
@@ -693,6 +682,17 @@ impl Chain {
 			"chain: txhashset_write: finished committing the batch (head etc.)"
 		);
 
+		// Replace the chain txhashset with the newly built one.
+		{
```
Comment (antiochp) [inline chain/src/chain.rs]: The concern here is that we have a (small) period of time between committing the batch and swapping the txhashset out for the new one.
The db/index is no longer consistent with the old txhashset at this point.
So if any other threads reads from the db we are at risk of reading inconsistent data.

I'm not sure how much of a problem this is in practice - need to think about it a bit more.

Comment (garyyu) [reply chain/src/chain.rs]: Or maybe move to exact front of `batch.commit()` ?  more safe?

https://github.com/mimblewimble/grin/blob/7b0c8d36fc9063aec82aef220aa73f1dc460a23f/chain/src/chain.rs#L678

Comment (antiochp) [reply chain/src/chain.rs]: There's still a non-atomic operation happening here, still a period when batch is committed but txhashet is not yet updated.

Comment (garyyu) [reply chain/src/chain.rs]: You want a mutex to link both txhashset write and batch commit ?

**Comment by antiochp** [comment]
We have a race condition either way - 
* update txhashset first, another thread could attempt to read it based on old indices
* commit batch first, another thread uses new indices to read the old txhashset

So I think we're fine making the change you propose (and keeping batch vs. txhashset ordering consistent everywhere).

:+1:

**Pull request merged**